### DISCUSSION TITLE
A newbie query on imagebuildah.BuildDockerfiles()

### DISCUSSION THREAD

**Discussion by santoshkal** [containers/buildah#5536] [General]
Hi everyone,

I'm new to Buildah and I'm trying to build a container from a Dockerfile using `imagebuildah.BuildDockerfiles()`. I've used the same `storage.Store` and `define.BuildOptions` as in the tests.

Here's a snippet of my code:

```go
func runBuildCmd(cmd *cobra.Command, args []string) error {
        // ...
	if buildah.InitReexec() {
		return fmt.Errorf("Error initializing: %v", cmd)
	}
	unshare.MaybeReexecUsingUserNamespace(false)
	// ...
        // var buildStore storage.Store
	buildStore, err := storage.GetStore(buildStoreOptions)
	if err != nil {
		panic(err)
	}
	defer func() {
		if _, err := buildStore.Shutdown(false); err != nil {
			if !errors.Is(err, storage.ErrLayerUsedByContainer) {
				fmt.Printf("Failed to shutdown storage: %q", err)
			}
		}
	}()
	// ...
	buildOptions := define.BuildOptions{
		ContextDirectory: buildArgs.reqinput,
		NamespaceOptions: []define.NamespaceOption{
			{Name: string(specs.NetworkNamespace), Host: true},
		},
	}

	ctx := context.TODO()
	id, ref, err := img.BuildDockerfiles(ctx, buildStore, buildOptions, dockerfile)
	if err != nil {
		return fmt.Errorf("Error building container from %s: %v", buildArgs.reqinput, err)
	}
	fmt.Printf("image %v created successfully with %v digest %v", id, ref)
	return nil
}
```
When I try to build this, I encounter the following error:
```shell
➜  genval git:(rego-results) ✗ go build -o ./bin/genval .
# github.com/containers/image/v5/storage
../go/pkg/mod/github.com/containers/image/v5@v5.30.1/storage/storage_dest.go:386:53: aLayer.UncompressedDigest undefined (type "github.com/containers/storage".AdditionalLayer has no field or method UncompressedDigest)
```

I am importing `github.com/containers/storage`. Can someone please help me understand what's going wrong?

Thank you!

**Comment by nalind** [comment]
This was not an error you made - the storage library changed an API that the image library was using.  If you temporarily force github.com/containers/storage down to v1.53.0, I think it'll clear up.

I realize now that it's not something that's explicitly said in the godoc for `buildah.InitReexec()`, but when `buildah.InitReexec()` retuns `true`, it's not an error, and `main()` should return with no error and an exit status of 0.  It's something you'd have run into eventually.

**Reply by santoshkal** [reply]
Thank you for your help; it helped resolve the initial errors. I apologize for the delayed response as I was on vacation.

While trying to follow `cmd/buildah/build.go` I am now facing a new issue when attempting to build this Dockerfile:
```Dockerfile
FROM alpine:latest
RUN echo "Hello, Buildah!"
```

Here is the error message:

```shell
TEP 1/2: FROM alpine:latest
Error: creating build container: could not find "netavark" in one of {[/usr/local/libexec/podman /usr/local/lib/podman /usr/libexec/podman /usr/lib/podman] {<nil>}}. To resolve this error, set the helper_binaries_dir key in the `[engine]` section of containers.conf to the directory containing your helper binaries.
```
I also looked at `tests/tutorial/tutorial.go` and observed the same behavior.
 
I'm wondering if this might be caused by some issues with setting up network-backend in the configuration in `define.BuildOptions`?
Is there something specific I should be looking for there?

Thanks again for your time and assistance!

**Reply by nalind** [reply]
The error message indicates that it's looking for `netavark`.  This check happens before the network configuration is evaluated, so you'll need to make sure it's installed.

**Reply by santoshkal** [reply]
Thank you.

Could you please direct me to resources that will help me install the `netavark` and resolve the errors?
I also examined `tests/tutorial/tutorial.go` and observed the same behavior.

**Reply by nalind** [reply]
Your OS distribution may have packages of it already, otherwise you'll need to build it from sources at https://github.com/containers/netavark.

**Reply by santoshkal** [reply]
I can build an image with `netavark` package installed. Thanks a lot for your help.

I had another query. Can I build an image from my Go build tool with any other network backend than `netavark`?

**Reply by nalind** [reply]
The github.com/containers/common library provides netavark and CNI.  But it's source code, so anything that implements the right Go interface should work, give or take bugs.
